### ISSUE TITLE
Original syntaxnode does not contain full parent syntax tree for return statement nodes

### ISSUE THREAD

**Issue by Mobrockers** [stryker-mutator/stryker-net#353]
This results in being unable to correctly report the mutation and original node location in source code as the line number is always 0 (as the return statement is the full syntax tree and the mutation is on the return statement value/expression).

**Comment by Mobrockers** [comment]
Traced the issue to here: https://github.com/stryker-mutator/stryker-net/blob/master/src/Stryker.Core/Stryker.Core/Mutants/MutantOrchestrator.cs#L129-L134

```
// No statement found yet, search deeper in the tree for statements to mutate
var children = currentNode.ChildNodes().ToList(); // Children have parent here
var childCopy = currentNode.TrackNodes(children); // Childcopy has lost it's parent
foreach (var child in children)
{
    var mutatedNode = Mutate(child);
    var originalNode = childCopy.GetCurrentNode(child); // So original node here has current node as parent, but not it's parent. Syntaxtree is broken.
```

**Comment by richardwerkman** [comment]
Do you have an piece of code available that results in a mutation with a wrong span? I could look into this if I have a reproduction project.

**Comment by Mobrockers** [comment]
Our integrationtest StringMagic.Demo method on the return true -> false and DummyMath.Add method on the checked(a + b) -> a + b both have this behaviour

**Comment by richardwerkman** [comment]
The object `currentNode` still has the parent node available. We need a way to return the copy node with the parent of the currentNode.

``` cs
// No statement found yet, search deeper in the tree for statements to mutate
var children = currentNode.ChildNodes().ToList();
var childCopy = currentNode.TrackNodes(children);
foreach (var child in children)
{
    var mutatedNode = Mutate(child);
    var originalNode = childCopy.GetCurrentNode(child);
    if (!mutatedNode.IsEquivalentTo(originalNode))
    {
         childCopy = childCopy.ReplaceNode(originalNode, mutatedNode);
    }
}
// currentNode still has the parent here.
// childCopy has parent null
// pseudocode:
// childCopy.parent = currentNode.parent
return childCopy;
```

**Comment by richardwerkman** [comment]
I have to review my last comment. It is not of importance that the node that is returned has a parent node. What is of importance is that the currentNode in `Mutate` always has a parent node. 

When does this not occur? If we call `replaceNode` on a tree and then call `Mutate` on that replaced tree. The method replaceNode removes the parent node of the tree. 

So we have to introduce a new thumb rule for the whole MutantOrchestrator: 

> `Mutate(SyntaxNode)` and `ApplyMutants(SyntaxNode)` may not be called on a node that results from `replaceNode` or `trackNodes`

If we apply this rule to all logic inside MutantOrchestrator it is assured that all mutations will have their parent nodes and all mutation spans will be correct in the report.

**Comment by dupdob** [comment]
PR #373 fixes that, I think. But I guess there is a need to have tests for that.

**Comment by Mobrockers** [comment]
@dupdob In #352 I will add checking on span locations in the integration test, so this will cover parents being available. If you feel like adding additional unit tests for it by all means, but it is not required.

**Comment by dupdob** [comment]
I added some assertions on mutant location in MutantOrchestratorTests

**Comment by Mobrockers** [comment]
Fixed by #373
